packages/ui/src/CodeEditor.tsx

```diff
@@ -0,0 +1,178 @@
+
+  for more see https://github.com/codemirror/legacy-modes
+
+  ## About styling/themes
+
+  CM6 is not that rich with predefined themes yet so we are not providing anything other than the
+  default theme (though it's easy provide custom ones if/when we need it.)
+
+  ## Customizing the editor
+
+  For further customizations, one can use .customExtensions prop to extend the editor with
+  any CM6 extension possible. (see Code component for an example.)
+
+  For getting a handle to the actual CM6 instance one may use the .innerRef prop which refers
+  to the EditorView instance created behind the scenes. (See the story of this component)
+*/
+
+import React, { FC, useEffect, useRef, useImperativeHandle, MutableRefObject } from 'react'
+import cn from 'classnames'
+import { EditorView, ViewUpdate, highlightSpecialChars, drawSelection, highlightActiveLine, keymap } from '@codemirror/view'
+import { EditorState, Extension } from '@codemirror/state'
+import { history, historyKeymap } from '@codemirror/history'
+import { foldGutter, foldKeymap } from '@codemirror/fold'
+import { indentOnInput } from '@codemirror/language'
+import { lineNumbers } from '@codemirror/gutter'
+import { defaultKeymap } from '@codemirror/commands'
+import { bracketMatching } from '@codemirror/matchbrackets'
+import { closeBrackets, closeBracketsKeymap } from '@codemirror/closebrackets'
+import { highlightSelectionMatches, searchKeymap } from '@codemirror/search'
+import { autocompletion, completionKeymap } from '@codemirror/autocomplete'
+import { commentKeymap } from '@codemirror/comment'
+import { rectangularSelection } from '@codemirror/rectangular-selection'
+import { defaultHighlightStyle } from '@codemirror/highlight'
+import { lintKeymap } from '@codemirror/lint'
+import { StreamLanguage } from '@codemirror/stream-parser'
+
+import { useDeepCompareMemo } from './hooks/useDeepCompareMemo'
+import styles from './CodeEditor.module.scss'
+
+// Export these very common CodeMirror types for ease of use
+export { EditorView, EditorState }
+
+// Common options for the component.
+// More advanced configuration of the underlying component should be done directly via a handle.
+export type CodeOptions = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  language?: any
+  lineWrapping?: boolean
+  lineNumbers?: boolean
+}
+
+export interface EditorViewRef {
+  view(): EditorView
+}
+
+export type CodeEditorProps = {
+  value?: string
+  className?: string
+  options?: CodeOptions
+  onChange?: (val: string) => void
+  customExtensions?: Extension[]
+  innerRef?: MutableRefObject<EditorViewRef | null>
+}
+
+const DEFAULT_OPTIONS: CodeOptions = {
+  language: null,
+  lineWrapping: false,
+  lineNumbers: true,
+}
+
+export const CodeEditor: FC<CodeEditorProps> = ({ value, className, options = {}, onChange, customExtensions, innerRef }) => {
+  const parentRef = useRef<HTMLDivElement>(null)
```

Comment:
```suggestion
  const parentRef = useRef<HTMLDivElement | null>(null)
```

like the line below?

Reply:
another nice catch! resolving.